File: packages/v2-fullstack/backend/app/schemas/shared.py
"""
Shared Contract: 统一维度、DTO分类、枚举定义

本模块定义了v2全栈架构的核心契约,确保前后端使用一致的:
- 输入维度命名与语义
- 输出DTO分类
- 时间口径
- Mode裁剪规则
- predicted批次一致性规则

Reference:
- docs/v2/v2实施细则/01-Shared-Contract基线-细则.md
- docs/v2/v2复用逻辑摘录/RD-共享类型与接口契约.md
"""

from datetime import datetime, timezone
from decimal import Decimal
from enum import Enum
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ============================================================================
# Enums (枚举类型 - 必须与前端保持一致)
# ============================================================================

class RegionScope(str, Enum):
    """区域范围级别"""
    PROVINCE = "province"
    DISTRICT = "district"
    # 可扩展: CITY = "city", GRID = "grid"


class DataType(str, Enum):
    """数据类型"""
    HISTORICAL = "historical"
    PREDICTED = "predicted"


class WeatherType(str, Enum):
    """天气类型 - 参考 RD-多天气类型扩展.md"""
    RAINFALL = "rainfall"
    WIND = "wind"
    TEMPERATURE = "temperature"
    # 未来扩展: HUMIDITY = "humidity", PRESSURE = "pressure"


class AccessMode(str, Enum):
    """访问模式 - 影响数据裁剪、默认展开、可用动作"""
    DEMO_PUBLIC = "demo_public"      # Demo/Public: 路演默认,少数字强可视化
    PARTNER = "partner"              # Partner: 合作伙伴,更深KPI与对比
    ADMIN_INTERNAL = "admin_internal"  # Admin/Internal: 内部,全量明细


# ============================================================================
# Input Dimensions (输入维度 - 最小集合)
# ============================================================================

class TimeRange(BaseModel):
    """
    时间范围
    
    规则:
    - 存储与传输使用UTC
    - 业务边界对齐使用region_timezone
    - 展示时区仅用于UI渲染
    
    Reference: RD-时间与时区口径统一.md
    """
    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "start": "2025-01-01T00:00:00Z",
                "end": "2025-01-31T23:59:59Z"
            }
        }
    )
    
    start: datetime = Field(
        ...,
        description="起始时间(UTC)",
    )
    end: datetime = Field(
        ...,
        description="结束时间(UTC)",
    )
    
    @field_validator("end")
    @classmethod
    def validate_end_after_start(cls, v: datetime, info) -> datetime:
        if "start" in info.data and v <= info.data["start"]:
            raise ValueError("end must be after start")
        return v


class SharedDimensions(BaseModel):
    """
    共享输入维度 - 所有Data Product请求的最小集合
    
    硬规则:
    - region_code/time_range/data_type/weather_type/access_mode 为必须
    - predicted场景下 prediction_run_id 为必须
    - 任何新增维度必须先入契约再入实现
    
    缓存key维度: 至少包含 access_mode; predicted 额外包含 prediction_run_id
    """
    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "region_scope": "province",
                "region_code": "CN-GD",
                "time_range": {
                    "start": "2025-01-01T00:00:00Z",
                    "end": "2025-01-31T23:59:59Z"
                },
                "data_type": "historical",
                "weather_type": "rainfall",
                "product_id": "daily_rainfall",
                "access_mode": "demo_public",
                "prediction_run_id": None
            }
        }
    )
    
    # 必须维度
    region_scope: RegionScope = Field(
        ...,
        description="区域范围级别"
    )
    region_code: str = Field(
        ...,
        description="区域代码(统一编码,如CN-GD表示广东省)",
        min_length=2,
        max_length=20
    )
    time_range: TimeRange = Field(
        ...,
        description="时间范围(UTC)"
    )
    data_type: DataType = Field(
        ...,
        description="数据类型"
    )
    weather_type: WeatherType = Field(
        ...,
        description="天气类型"
    )
    access_mode: AccessMode = Field(
        ...,
        description="访问模式(影响输出裁剪)"
    )
    
    # 可选维度
    product_id: Optional[str] = Field(
        None,
        description="产品ID(可选,但一旦使用必须入缓存key)"
    )
    prediction_run_id: Optional[str] = Field(
        None,
        description="预测批次ID(predicted场景必须)"
    )
    region_timezone: Optional[str] = Field(
        None,
        description="区域时区(如Asia/Shanghai,用于业务边界对齐)",
        examples=["Asia/Shanghai", "America/New_York"]
    )
    
    @field_validator("prediction_run_id")
    @classmethod
    def validate_prediction_run_id(cls, v: Optional[str], info) -> Optional[str]:
        """predicted场景下必须提供prediction_run_id"""
        if "data_type" in info.data:
            data_type = info.data["data_type"]
            if data_type == DataType.PREDICTED and not v:
                raise ValueError("prediction_run_id is required when data_type is predicted")
            if data_type == DataType.HISTORICAL and v:
                raise ValueError("prediction_run_id must be None when data_type is historical")
        return v
    
    def to_cache_key(self) -> str:
        """
        生成缓存key
        
        硬规则:
        - 至少包含: region_scope, region_code, time_range, data_type, weather_type, access_mode
        - predicted场景额外包含: prediction_run_id
        """
        parts = [
            f"region:{self.region_scope.value}:{self.region_code}",
            f"time:{self.time_range.start.isoformat()}:{self.time_range.end.isoformat()}",
            f"dtype:{self.data_type.value}",
            f"weather:{self.weather_type.value}",
            f"mode:{self.access_mode.value}",
        ]
        
        if self.product_id:
            parts.append(f"product:{self.product_id}")
        
        if self.data_type == DataType.PREDICTED and self.prediction_run_id:
            parts.append(f"run:{self.prediction_run_id}")
        
        return "|".join(parts)


# ============================================================================
# Output DTOs (输出DTO分类 - 避免混用)
# ============================================================================

class SeriesData(BaseModel):
    """
    时间序列数据
    
    用途: Timeline三泳道(Weather/Risk/Claims)、趋势图
    """
    model_config = ConfigDict(from_attributes=True)
    
    timestamps: List[datetime] = Field(
        ...,
        description="时间点列表(UTC)"
    )
    values: List[Union[float, Decimal, int]] = Field(
        ...,
        description="数值列表"
    )
    unit: str = Field(
        ...,
        description="单位(如mm, km/h, celsius)",
        examples=["mm", "km/h", "celsius", "CNY"]
    )


class EventData(BaseModel):
    """
    事件数据
    
    用途: risk_events, claim_events
    
    硬规则:
    - predicted事件必须包含 prediction_run_id
    - 必须包含 rule_version 或 rules_hash (支持审计与回溯)
    """
    model_config = ConfigDict(from_attributes=True)
    
    event_id: str = Field(..., description="事件ID")
    timestamp: datetime = Field(..., description="事件时间(UTC)")
    event_type: str = Field(..., description="事件类型(risk/claim)")
    
    # 可选字段(根据事件类型动态填充)
    tier_level: Optional[int] = Field(None, description="风险等级(1/2/3)")
    trigger_value: Optional[Decimal] = Field(None, description="触发值")
    threshold_value: Optional[Decimal] = Field(None, description="阈值")
    amount: Optional[Decimal] = Field(None, description="金额(理赔)")
    
    # 审计字段
    data_type: DataType = Field(..., description="数据类型")
    prediction_run_id: Optional[str] = Field(
        None,
        description="预测批次ID(predicted事件必须)"
    )
    rule_version: Optional[str] = Field(
        None,
        description="规则版本(支持审计)"
    )


class AggregationData(BaseModel):
    """
    聚合数据
    
    用途: L0 KPI/Pareto（TopN）, Map Overlays 区域聚合
    
    硬规则:
    - 必须标注聚合维度(aggregation_key)与口径(aggregation_method)
    """
    model_config = ConfigDict(from_attributes=True)
    
    aggregation_key: str = Field(
        ...,
        description="聚合维度(如region_code, product_id)"
    )
    aggregation_method: str = Field(
        ...,
        description="聚合方法(sum/avg/count/max/min)",
        examples=["sum", "avg", "count", "max", "min"]
    )
    value: Union[float, Decimal, int] = Field(
        ...,
        description="聚合值"
    )
    unit: Optional[str] = Field(
        None,
        description="单位(可选)"
    )
    label: Optional[str] = Field(
        None,
        description="展示标签(可选)"
    )


class LegendMeta(BaseModel):
    """
    图例与元信息
    
    用途: 解释数据口径、单位、阈值、tiers等
    
    硬规则:
    - predicted数据必须包含 prediction_run_id
    - 必须标注 data_type 与 weather_type
    """
    model_config = ConfigDict(from_attributes=True)
    
    data_type: DataType = Field(..., description="数据类型")
    weather_type: WeatherType = Field(..., description="天气类型")
    unit: str = Field(..., description="单位")
    
    # 阈值/tiers(可选)
    thresholds: Optional[Dict[str, Union[float, Decimal]]] = Field(
        None,
        description="阈值字典(如 {'tier1': 50, 'tier2': 100, 'tier3': 150})",
        examples=[{"tier1": 50.0, "tier2": 100.0, "tier3": 150.0}]
    )
    
    # predicted元信息
    prediction_run_id: Optional[str] = Field(
        None,
        description="预测批次ID(predicted必须)"
    )
    prediction_generated_at: Optional[datetime] = Field(
        None,
        description="预测生成时间(UTC)"
    )
    
    # 口径说明(Mode-aware)
    description: Optional[str] = Field(
        None,
        description="口径说明(根据access_mode裁剪)"
    )


class DataProductResponse(BaseModel):
    """
    Data Product统一响应格式
    
    包含:
    - series: 时间序列(可选)
    - events: 事件列表(可选)
    - aggregations: 聚合数据(可选)
    - legend: 图例与元信息(必须)
    - meta: 响应元数据(可观测性)
    """
    model_config = ConfigDict(from_attributes=True)
    
    series: Optional[List[SeriesData]] = Field(
        None,
        description="时间序列数据"
    )
    events: Optional[List[EventData]] = Field(
        None,
        description="事件数据"
    )
    aggregations: Optional[List[AggregationData]] = Field(
        None,
        description="聚合数据"
    )
    legend: LegendMeta = Field(
        ...,
        description="图例与元信息"
    )
    meta: "ResponseMeta" = Field(
        ...,
        description="响应元数据"
    )


# ============================================================================
# Observability (可观测性)
# ============================================================================

class TraceContext(BaseModel):
    """
    追踪上下文 - 支撑全链路可追溯
    
    必带字段:
    - trace_id/correlation_id
    - 关键维度(access_mode, region_code, data_type, weather_type, product_id, prediction_run_id)
    """
    model_config = ConfigDict(from_attributes=True)
    
    trace_id: str = Field(
        ...,
        description="追踪ID(全链路唯一)"
    )
    correlation_id: Optional[str] = Field(
        None,
        description="关联ID(可选,用于跨服务关联)"
    )
    
    # 关键维度(用于排障)
    access_mode: AccessMode
    region_code: Optional[str] = None
    time_range_start: Optional[datetime] = None
    time_range_end: Optional[datetime] = None
    data_type: Optional[DataType] = None
    weather_type: Optional[WeatherType] = None
    product_id: Optional[str] = None
    prediction_run_id: Optional[str] = None
    
    # 时间戳
    request_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        description="请求时间(UTC)"
    )


class ResponseMeta(BaseModel):
    """
    响应元数据
    
    用途: 可观测性、缓存状态、警告信息
    """
    model_config = ConfigDict(from_attributes=True)
    
    trace_context: TraceContext = Field(
        ...,
        description="追踪上下文"
    )
    cached: bool = Field(
        default=False,
        description="是否命中缓存"
    )
    cache_key: Optional[str] = Field(
        None,
        description="缓存key(用于排障)"
    )
    warnings: Optional[List[str]] = Field(
        None,
        description="警告信息(如数据不完整、降级处理等)"
    )
    response_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        description="响应时间(UTC)"
    )


# Forward reference resolution
DataProductResponse.model_rebuild()

File: packages/v2-fullstack/frontend/src/types/shared.ts
/**
 * Shared Contract: 统一维度、DTO分类、枚举定义
 * 
 * 本模块定义了v2全栈架构的核心契约,确保前后端使用一致的:
 * - 输入维度命名与语义
 * - 输出DTO分类
 * - 时间口径
 * - Mode裁剪规则
 * - predicted批次一致性规则
 * 
 * Reference:
 * - docs/v2/v2实施细则/01-Shared-Contract基线-细则.md
 * - docs/v2/v2复用逻辑摘录/RD-共享类型与接口契约.md
 * 
 * CRITICAL: 必须与 backend/app/schemas/shared.py 保持完全一致
 */

// ============================================================================
// Enums (枚举类型 - 必须与后端保持一致)
// ============================================================================

/**
 * 区域范围级别
 */
export enum RegionScope {
  PROVINCE = 'province',
  DISTRICT = 'district',
  // 可扩展: CITY = 'city', GRID = 'grid'
}

/**
 * 数据类型
 */
export enum DataType {
  HISTORICAL = 'historical',
  PREDICTED = 'predicted',
}

/**
 * 天气类型
 * Reference: RD-多天气类型扩展.md
 */
export enum WeatherType {
  RAINFALL = 'rainfall',
  WIND = 'wind',
  TEMPERATURE = 'temperature',
  // 未来扩展: HUMIDITY = 'humidity', PRESSURE = 'pressure'
}

/**
 * 访问模式 - 影响数据裁剪、默认展开、可用动作
 */
export enum AccessMode {
  DEMO_PUBLIC = 'demo_public',      // Demo/Public: 路演默认,少数字强可视化
  PARTNER = 'partner',              // Partner: 合作伙伴,更深KPI与对比
  ADMIN_INTERNAL = 'admin_internal',  // Admin/Internal: 内部,全量明细
}

// ============================================================================
// Input Dimensions (输入维度 - 最小集合)
// ============================================================================

/**
 * 时间范围
 * 
 * 规则:
 * - 存储与传输使用UTC
 * - 业务边界对齐使用region_timezone
 * - 展示时区仅用于UI渲染
 * 
 * Reference: RD-时间与时区口径统一.md
 */
export interface TimeRange {
  /** 起始时间(UTC) */
  start: string; // ISO 8601 format
  /** 结束时间(UTC) */
  end: string;   // ISO 8601 format
}

/**
 * 共享输入维度 - 所有Data Product请求的最小集合
 * 
 * 硬规则:
 * - region_code/time_range/data_type/weather_type/access_mode 为必须
 * - predicted场景下 prediction_run_id 为必须
 * - 任何新增维度必须先入契约再入实现
 * 
 * 缓存key维度: 至少包含 access_mode; predicted 额外包含 prediction_run_id
 */
export interface SharedDimensions {
  // 必须维度
  /** 区域范围级别 */
  region_scope: RegionScope;
  /** 区域代码(统一编码,如CN-GD表示广东省) */
  region_code: string;
  /** 时间范围(UTC) */
  time_range: TimeRange;
  /** 数据类型 */
  data_type: DataType;
  /** 天气类型 */
  weather_type: WeatherType;
  /** 访问模式(影响输出裁剪) */
  access_mode: AccessMode;
  
  // 可选维度
  /** 产品ID(可选,但一旦使用必须入缓存key) */
  product_id?: string;
  /** 预测批次ID(predicted场景必须) */
  prediction_run_id?: string;
  /** 区域时区(如Asia/Shanghai,用于业务边界对齐) */
  region_timezone?: string;
}

/**
 * 生成缓存key
 * 
 * 硬规则:
 * - 至少包含: region_scope, region_code, time_range, data_type, weather_type, access_mode
 * - predicted场景额外包含: prediction_run_id
 */
export function toCacheKey(dimensions: SharedDimensions): string {
  const parts = [
    `region:${dimensions.region_scope}:${dimensions.region_code}`,
    `time:${dimensions.time_range.start}:${dimensions.time_range.end}`,
    `dtype:${dimensions.data_type}`,
    `weather:${dimensions.weather_type}`,
    `mode:${dimensions.access_mode}`,
  ];
  
  if (dimensions.product_id) {
    parts.push(`product:${dimensions.product_id}`);
  }
  
  if (dimensions.data_type === DataType.PREDICTED && dimensions.prediction_run_id) {
    parts.push(`run:${dimensions.prediction_run_id}`);
  }
  
  return parts.join('|');
}

// ============================================================================
// Output DTOs (输出DTO分类 - 避免混用)
// ============================================================================

/**
 * 时间序列数据
 * 
 * 用途: Timeline三泳道(Weather/Risk/Claims)、趋势图
 */
export interface SeriesData {
  /** 时间点列表(UTC, ISO 8601 format) */
  timestamps: string[];
  /** 数值列表 */
  values: Array<number | string>; // string for Decimal support
  /** 单位(如mm, km/h, celsius) */
  unit: string;
}

/**
 * 事件数据
 * 
 * 用途: risk_events, claim_events
 * 
 * 硬规则:
 * - predicted事件必须包含 prediction_run_id
 * - 必须包含 rule_version 或 rules_hash (支持审计与回溯)
 */
export interface EventData {
  /** 事件ID */
  event_id: string;
  /** 事件时间(UTC, ISO 8601 format) */
  timestamp: string;
  /** 事件类型(risk/claim) */
  event_type: string;
  
  // 可选字段(根据事件类型动态填充)
  /** 风险等级(1/2/3) */
  tier_level?: number;
  /** 触发值 */
  trigger_value?: number | string; // string for Decimal
  /** 阈值 */
  threshold_value?: number | string;
  /** 金额(理赔) */
  amount?: number | string;
  
  // 审计字段
  /** 数据类型 */
  data_type: DataType;
  /** 预测批次ID(predicted事件必须) */
  prediction_run_id?: string;
  /** 规则版本(支持审计) */
  rule_version?: string;
}

/**
 * 聚合数据
 * 
 * 用途: L0 KPI/Pareto（TopN）, Map Overlays区域聚合
 * 
 * 硬规则:
 * - 必须标注聚合维度(aggregation_key)与口径(aggregation_method)
 */
export interface AggregationData {
  /** 聚合维度(如region_code, product_id) */
  aggregation_key: string;
  /** 聚合方法(sum/avg/count/max/min) */
  aggregation_method: 'sum' | 'avg' | 'count' | 'max' | 'min';
  /** 聚合值 */
  value: number | string; // string for Decimal
  /** 单位(可选) */
  unit?: string;
  /** 展示标签(可选) */
  label?: string;
}

/**
 * 图例与元信息
 * 
 * 用途: 解释数据口径、单位、阈值、tiers等
 * 
 * 硬规则:
 * - predicted数据必须包含 prediction_run_id
 * - 必须标注 data_type 与 weather_type
 */
export interface LegendMeta {
  /** 数据类型 */
  data_type: DataType;
  /** 天气类型 */
  weather_type: WeatherType;
  /** 单位 */
  unit: string;
  
  // 阈值/tiers(可选)
  /** 阈值字典(如 {'tier1': 50, 'tier2': 100, 'tier3': 150}) */
  thresholds?: Record<string, number | string>;
  
  // predicted元信息
  /** 预测批次ID(predicted必须) */
  prediction_run_id?: string;
  /** 预测生成时间(UTC, ISO 8601 format) */
  prediction_generated_at?: string;
  
  // 口径说明(Mode-aware)
  /** 口径说明(根据access_mode裁剪) */
  description?: string;
}

/**
 * Data Product统一响应格式
 * 
 * 包含:
 * - series: 时间序列(可选)
 * - events: 事件列表(可选)
 * - aggregations: 聚合数据(可选)
 * - legend: 图例与元信息(必须)
 * - meta: 响应元数据(可观测性)
 */
export interface DataProductResponse {
  /** 时间序列数据 */
  series?: SeriesData[];
  /** 事件数据 */
  events?: EventData[];
  /** 聚合数据 */
  aggregations?: AggregationData[];
  /** 图例与元信息 */
  legend: LegendMeta;
  /** 响应元数据 */
  meta: ResponseMeta;
}

// ============================================================================
// Observability (可观测性)
// ============================================================================

/**
 * 追踪上下文 - 支撑全链路可追溯
 * 
 * 必带字段:
 * - trace_id/correlation_id
 * - 关键维度(access_mode, region_code, data_type, weather_type, product_id, prediction_run_id)
 */
export interface TraceContext {
  /** 追踪ID(全链路唯一) */
  trace_id: string;
  /** 关联ID(可选,用于跨服务关联) */
  correlation_id?: string;
  
  // 关键维度(用于排障)
  access_mode: AccessMode;
  region_code?: string;
  time_range_start?: string; // ISO 8601 format
  time_range_end?: string;
  data_type?: DataType;
  weather_type?: WeatherType;
  product_id?: string;
  prediction_run_id?: string;
  
  // 时间戳
  /** 请求时间(UTC, ISO 8601 format) */
  request_at: string;
}

/**
 * 响应元数据
 * 
 * 用途: 可观测性、缓存状态、警告信息
 */
export interface ResponseMeta {
  /** 追踪上下文 */
  trace_context: TraceContext;
  /** 是否命中缓存 */
  cached: boolean;
  /** 缓存key(用于排障) */
  cache_key?: string;
  /** 警告信息(如数据不完整、降级处理等) */
  warnings?: string[];
  /** 响应时间(UTC, ISO 8601 format) */
  response_at: string;
}

// ============================================================================
// Validation Helpers
// ============================================================================

/**
 * 验证SharedDimensions
 * 
 * 硬规则:
 * - predicted场景下必须提供prediction_run_id
 * - historical场景下不能提供prediction_run_id
 */
export function validateSharedDimensions(dimensions: SharedDimensions): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];
  
  // 验证prediction_run_id
  if (dimensions.data_type === DataType.PREDICTED && !dimensions.prediction_run_id) {
    errors.push('prediction_run_id is required when data_type is predicted');
  }
  
  if (dimensions.data_type === DataType.HISTORICAL && dimensions.prediction_run_id) {
    errors.push('prediction_run_id must not be provided when data_type is historical');
  }
  
  // 验证time_range
  const start = new Date(dimensions.time_range.start);
  const end = new Date(dimensions.time_range.end);
  if (end <= start) {
    errors.push('time_range.end must be after time_range.start');
  }
  
  // 验证region_code
  if (dimensions.region_code.length < 2 || dimensions.region_code.length > 20) {
    errors.push('region_code length must be between 2 and 20');
  }
  
  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * 创建默认TraceContext
 */
export function createTraceContext(
  dimensions: Partial<SharedDimensions>,
  traceId?: string
): TraceContext {
  return {
    trace_id: traceId || `trace-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
    access_mode: dimensions.access_mode || AccessMode.DEMO_PUBLIC,
    region_code: dimensions.region_code,
    time_range_start: dimensions.time_range?.start,
    time_range_end: dimensions.time_range?.end,
    data_type: dimensions.data_type,
    weather_type: dimensions.weather_type,
    product_id: dimensions.product_id,
    prediction_run_id: dimensions.prediction_run_id,
    request_at: new Date().toISOString(),
  };
}
